### PR TITLE
fix(es/minifier): Bailout Regex optimization when facing invalid flags

### PR DESCRIPTION
**Pull request by Austaras**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

**Review by kdy1 — DISMISSED** [review]
Thank you!

---
swc-bump:
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs [hunk 1]
```diff
@@ -247,76 +248,73 @@ impl Pure<'_> {
...
+                Some(s.value.clone())
+            } else {
+                None
+            }
         }
+        fn valid_flag(flag: &Expr) -> Option<JsWord> {
+            if let Expr::Lit(Lit::Str(s)) = flag {
+                let mut set = FxHashSet::default();
+                for c in s.value.chars() {
+                    if !matches!(c, 'd' | 'g' | 'i' | 'm' | 's' | 'u' | 'y') {
```
Comment (magic-akari) [inline crates/swc_ecma_minifier/src/compress/pure/misc.rs:272]: These flags are considered "invalid" in the absence of a polyfill, specifically regarding the sticky and unicode features. check the core-js-compat data.

```json
{
  "es.regexp.sticky": {
    "android": "49",
    "bun": "0.1.1",
    "chrome": "49",
    "chrome-android": "49",
    "deno": "1.0",
    "edge": "13",
    "electron": "0.37",
    "firefox": "3",
    "firefox-android": "4",
    "hermes": "0.3",
    "ios": "10.0",
    "node": "6.0",
    "oculus": "3.0",
    "opera": "36",
    "opera-android": "36",
    "opera_mobile": "36",
    "quest": "3.0",
    "react-native": "0.69",
    "safari": "10.0",
    "samsung": "5.0"
  }
}
```

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs [hunk 2]
```diff
@@ -247,76 +248,78 @@ impl Pure<'_> {
...
     /// `new RegExp("([Sap]+)", "ig")` => `/([Sap]+)/gi`
     fn optimize_regex(&mut self, args: &mut Vec<ExprOrSpread>, span: &mut Span) -> Option<Expr> {
-        if args.is_empty() || args.len() > 2 {
-            return None;
+        fn valid_pattern(pattern: &Expr) -> Option<JsWord> {
+            if let Expr::Lit(Lit::Str(s)) = pattern {
+                if s.value.contains(|c: char| {
+                    // whitelist
+                    !c.is_ascii_alphanumeric()
+                        && !matches!(c, '$' | '[' | ']' | '(' | ')' | '{' | '}' | '-' | '+' | '_')
```
Comment (jridgewell) [inline crates/swc_ecma_minifier/src/compress/pure/misc.rs:256]: Just making note of the bug fix here: `%` -> `$`

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs [hunk 3]
```diff
@@ -247,76 +248,78 @@ impl Pure<'_> {
...
+                }) {
+                    return None;
+                }
+                if s.value.contains("\\\0") || s.value.contains('/') {
```
Comment (jridgewell) [inline crates/swc_ecma_minifier/src/compress/pure/misc.rs:260]: Nit: I think this is redundant? I think both the `\0` null byte and `/` will be caught be the above `!is_ascii_alphanumeric() && !matches()`?

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs [hunk 4]
```diff
@@ -247,76 +248,78 @@ impl Pure<'_> {
...
+            } else {
+                None
+            }
         }
+        fn valid_flag(flag: &Expr, es_version: EsVersion) -> Option<JsWord> {
+            if let Expr::Lit(Lit::Str(s)) = flag {
+                let mut set = FxHashSet::default();
+                for c in s.value.chars() {
+                    if !(matches!(c, 'd' | 'g' | 'i' | 'm')
+                        || (es_version >= EsVersion::Es2015 && matches!(c, 's' | 'u' | 'y')))
```
[the remark this answers is not in the document]
Comment (Austaras) [reply crates/swc_ecma_minifier/src/compress/pure/misc.rs:273]: I check again so `gim` is included from the beginning, `uy` is introduced in [es6](https://262.ecma-international.org/6.0/#sec-get-regexp.prototype.flags), `s` in [es2018](https://github.com/tc39/proposal-regexp-dotall-flag) and `d` in [es2022](https://github.com/tc39/proposal-regexp-match-indices)

[the remark this answers is not in the document]
Comment (jridgewell) [reply crates/swc_ecma_minifier/src/compress/pure/misc.rs:273]: Ah, sorry, I got the 2015 and 2018 backwards.

**Review by jridgewell — APPROVED** [review]

**Pull request merged**